### ISSUE TITLE
Feature Request: matrix builds allowed failures

### ISSUE THREAD

**Issue by patrickjahns** [drone/drone#1920]
It would be great to extend matrix builds, to allow for failure of certain combinations.

Use case:
We are developing against certain baselines of libraries/frameworks - very often the 'on the edge versions' are not yet supported. Yet it would be great to include them already in the CI workflow to see early on, when upcoming versions of a library/framework/language would result in breaking changes. 

However - these versions should not fail the whole build, since they are not required to pass.

We could add another param to matrix builds called `allowed_failures` (similar to [the way travis handles it](https://docs.travis-ci.com/user/customizing-the-build#Rows-that-are-Allowed-to-Fail))

Example .drone.yml
```
pipeline:
  build: 
    image: myframeworkimage:${VERSION}

matrix:
  VERSION:
    - 1 
    - 2
  allowed_failures:
    - VERSION: 1
```

I`ve looked through the code but am not quite sure if the workflow of the current build steps/jobs would allow for this. Any hints/ideas for where to dig deeper in the source are welcome - will see what I can do

**Comment by bradrydzewski** [comment]
I come from the school of thought that if it is worth complaining about, it is worth fixing. What is the value of running continuous tests against configurations you don't expect to pass?

**Comment by gtaylor** [comment]
Some projects show a full matrix of builds, some of which are running experimental or pre-release builds against certain architectures or compiler/interpreter versions. These absolutely don't represent a failure of the fully supported versions or architectures, but are something they want to keep an eye on.

**Comment by patrickjahns** [comment]
@bradrydzewski 
In our case, we develop extensions/plugins for vendor specific software. While we can control the versions we run for customers - we`d still like to keep an eye on early development from the vendor to see when breaking changes occur. When these changes occur, it doesn't have consequences yet for the current maintained versions - but knowing early on when braking changes occur helps to maintain these projects.

**Comment by bradrydzewski** [comment]
I understand the use case, I just want to make sure people use this in practice.  I could see a project enabling a bunch of matrix with failures, but never taking the time to fix, and then two years later they are still failing. This is based on my experience at some dysfunctional big corps, which I realize is not representative of the overall developer community, but it is still my job to play devils advocate and make sure we are adding useful features to Drone :)

If this is a critical part of your teams workflow, and you think this feature belongs in Drone I'm not going to stand in the way. It isn't that difficult to add, overall.

I would recommend chunking this into a few different pull requests.

__First__, we need to add a new field to the jobs database table and the jobs struct to label the job as one that can fail.

```diff
type Job struct {
	ID       int64  `json:"id"           meddler:"job_id,pk"`
	BuildID  int64  `json:"-"            meddler:"job_build_id"`
	NodeID   int64  `json:"-"            meddler:"job_node_id"`
	Number   int    `json:"number"       meddler:"job_number"`
	Error    string `json:"error"        meddler:"job_error"`
	Status   string `json:"status"       meddler:"job_status"`
	ExitCode int    `json:"exit_code"    meddler:"job_exit_code"`
+	AllowErr bool   `json:"allow_error"  meddler:"job_allow_error"`
	Enqueued int64  `json:"enqueued_at"  meddler:"job_enqueued"`
	Started  int64  `json:"started_at"   meddler:"job_started"`
	Finished int64  `json:"finished_at"  meddler:"job_finished"`

	Environment map[string]string `json:"environment" meddler:"job_environment,json"`
}
```

__Second__, we need to update the logic that updates the build status. It currently iterates through the list of all job statuses, and any non success statuses bubble up to the build. This logic would be updated to take `AllowErr` into account

```diff
	for _, job := range jobs {
+		if job.AllowErr {
+			continue
+		}
		if job.Finished > finish {
			finish = job.Finished
		}
		switch job.Status {
		case model.StatusSuccess:
			// no-op
		case model.StatusRunning, model.StatusPending:
			return false, nil
		default:
			status = job.Status
		}
	}
```

https://github.com/drone/drone/blob/master/store/store.go#L366:L377

__Third__, we would need to update the yaml parser. This is something that can be prototyped, but is not something we'll be able to merge for a couple weeks because this section of code is being overhauled as we speak, and moved to https://github.com/cncd

**Comment by patrickjahns** [comment]
@bradrydzewski 

Thank you for the hints - I will already prepare what I can do and open a PR for this for review.

For the yaml parser, I was already looking at:
https://github.com/drone/drone/blob/master/yaml/matrix.go

What is the reason for moving it to a different organisation alltogether?

**Comment by bradrydzewski** [comment]
> What is the reason for moving it to a different organization alltogether?

The answer to this question is a bit complex and I'm planning a detailed blog post. I am hoping to create a standards group that will focus on a common runtime for container pipelines (this will be a low level specification) and then a separate specification for a common yaml format, which would compile down to the common runtime. It will also demonstrate how the travis, bitbucket and gitlab yaml files can compile down to the common runtime.

**Comment by jmccann** [comment]
Can I verify that for this feature we are looking to allow a step to fail but continue the Drone run?  Should the title be renamed to "Feature Request: build job/step allowed failures"?

Also, if the above is true (that we are more generically looking to allow a step to fail) I'm very interested in this.  😄

**Comment by bradrydzewski** [comment]
@jmccann for individual steps we have https://github.com/drone/drone/issues/2071. This has been implemented in the yaml and runtime libraries [1][2] which will be available starting with 0.9 😄

[1] https://github.com/drone/drone-runtime/blob/master/engine/spec.go#L46
[2] https://github.com/drone/drone-yaml-v1/blob/master/yaml/container.go#L26

**Comment by bradrydzewski** [comment]
drone now supports `failure: ignore` in the step definition 

```
kind: pipeline
name: default

steps:
- name: build
  image: golang
  failure: ignore
  commands:
  - go build
  - go test -short
```
